### ISSUE TITLE
SLEAP 1.4.1 installation error on Windows - TensorFlow 2.7.0 unavailable/corrupted package archive

### ISSUE THREAD

**Issue by GarrettBlair** [talmolab/sleap#2244]
I'm upgrading my SLEAP install from v1.3.1 to 1.4.1, and have an error on install following the release instructions. Similar to <a href="https://github.com/talmolab/sleap/issues/2126">Issue 2126</a>. Transparency- I used ChatGPT to debug the installation issue, develop this fix, and format the bug report.
<p><strong>tl;dr:</strong><br>
SLEAP 1.4.1 cannot be installed on Windows via conda/mamba or pip with Python ≥3.9 because:</p>
<ol>
<li>Conda/mamba fails — TensorFlow 2.7.0 is not available for Windows on conda-forge/anaconda, and the sleap channel’s TF 2.7.0 build is broken.</li>
<li>Pip fails — PySide2 5.13–5.14 wheels don’t exist for Python ≥3.9.</li>
</ol>
<p><strong>Workaround:</strong> Use Python 3.8 and install via pip:</p>
<pre><code>mamba create -n sleap python=3.8
mamba activate sleap
pip install sleap[pypi]==1.4.1
</code></pre>
<p><strong>Suggestion:</strong> Update dependency pins or provide Windows-compatible wheels for TF 2.7.0 and PySide2 to support Python ≥3.9.</p>

<h1>Bug description</h1>
<p>Attempting to install <strong>SLEAP 1.4.1</strong> on Windows fails due to dependency resolution issues with TensorFlow 2.7.0 and PySide2 version pins. The environment cannot be solved via conda/mamba, and pip install fails for Python ≥3.9.</p>

<h2>Expected behaviour</h2>
<ul>
<li>Conda/mamba or pip should install SLEAP 1.4.1 on Windows successfully.</li>
<li>GUI should be functional post-installation.</li>
</ul>

<h2>Actual behaviour</h2>
<ul>
<li>Conda/mamba installation fails:
<pre><code>Could not solve for environment specs — no viable option for tensorflow==2.7.0</code></pre>
</li>
<li>Pip installation fails on Python ≥3.9:
<pre><code>No matching distribution found for PySide2&lt;=5.14.1, &gt;=5.13.2</code></pre>
</li>
<li>Workaround is to use Python 3.8.</li>
</ul>

<h2>Your personal set up</h2>
<ul>
<li>OS: Windows 11 (x86_64)</li>
<li>Version(s): SLEAP 1.4.1, Python 3.8 (working), Python ≥3.9 (fails)</li>
<li>SLEAP installation method:
<ul>
<li>[x] pip package (working workaround with Python 3.8)</li>
<li>[ ] Conda from package</li>
<li>[ ] Conda from source</li>
<li>[ ] Apple Silicon Macs</li>
</ul>
</li>
</ul>

<details><summary>Environment packages</summary>
<pre><code># Example of working environment
mamba create -n sleap python=3.8
mamba activate sleap
pip install sleap[pypi]==1.4.1

# pip freeze output
absl-py==1.4.0
accessible-pygments==0.0.4
alabaster==0.7.13
albumentations==1.3.1
anyio==3.7.0
appdirs==1.4.4
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
astunparse==1.6.3
attrs==21.4.0
Babel==2.12.1
backcall==0.2.0
backports.zoneinfo==0.2.1
beautifulsoup4==4.12.2
black==21.6b0
bleach==6.0.0
cached-property @ file:///tmp/build/80754af9/cached-property_1600785575025/work
cachetools==5.3.1
cattrs==1.1.1
certifi==2021.10.8
cffi==1.15.1
cfgv==3.3.1
charset-normalizer==3.1.0
click==8.0.4
colorama==0.4.6
commonmark==0.9.1
coverage==7.2.7
cryptography==41.0.1
cycler==0.11.0
debugpy==1.6.7
decorator==5.1.1
defusedxml==0.7.1
Deprecated==1.2.14
distlib==0.3.6
docutils==0.17.1
efficientnet==1.0.0
entrypoints==0.4
exceptiongroup==1.1.1
filelock==3.12.2
flatbuffers==23.5.26
fonttools==4.38.0
gast==0.5.4
gitdb==4.0.10
GitPython==3.1.31
google-auth==2.20.0
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
greenlet==2.0.2
grpcio==1.56.0
h5py @ file:///D:/bld/h5py_1604753757907/work
hdmf==3.5.5
identify==2.5.24
idna==3.4
image-classifiers==1.0.0
imageio==2.15.0
imageio-ffmpeg==0.5.1
imagesize==1.4.1
imgaug==0.4.0
imgstore==0.2.9
importlib-metadata==6.7.0
importlib-resources==5.12.0
iniconfig==2.0.0
ipykernel==6.16.2
ipython==7.34.0
ipython-genutils==0.2.0
ipywidgets==7.7.5
jaraco.classes==3.2.3
jedi==0.17.2
Jinja2==3.1.2
joblib==1.2.0
jsmin==3.0.1
json5==0.9.14
jsonpickle==1.2
jsonschema==4.17.3
jupyter-cache==0.4.3
jupyter-server==1.13.5
jupyter-server-mathjax==0.2.6
jupyter-sphinx==0.3.2
jupyter_client==7.4.9
jupyter_core==4.12.0
jupyterlab==3.3.4
jupyterlab-pygments==0.2.2
jupyterlab-widgets==1.1.4
jupyterlab_server==2.16.3
keras==2.8.0
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.2
keyring==24.1.0
kiwisolver==1.4.4
libclang==16.0.0
linkify-it-py==2.0.2
livereload==2.6.3
lxml==4.9.2
Markdown==3.4.3
markdown-it-py==1.1.0
MarkupSafe==2.1.3
matplotlib==3.5.3
matplotlib-inline==0.1.6
mdit-py-plugins==0.2.8
mistune==0.8.4
more-itertools==9.1.0
mypy-extensions==1.0.0
myst-nb==0.13.2
myst-parser==0.15.2
nbclassic==0.5.6
nbclient==0.5.13
nbconvert==6.5.4
nbdime==3.2.1
nbformat==5.1.3
ndx-pose==0.1.1
nest-asyncio==1.5.6
networkx==2.6.3
nixio==1.5.3
nodeenv==1.8.0
notebook==6.5.4
notebook_shim==0.2.3
numpy @ file:///D:/bld/numpy_1649585408897/work
oauthlib==3.2.2
olefile==0.46
opencv-python==4.5.5.64
opencv-python-headless==4.12.0.88
opt-einsum==3.3.0
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1681337016113/work
pandas @ file:///D:/bld/pandas_1639398349358/work
pandocfilters==1.5.0
parso==0.7.1
pathspec==0.11.1
pickleshare==0.7.5
Pillow @ file:///D:/bld/pillow_1636559064986/work
pkginfo==1.9.6
pkgutil_resolve_name==1.3.10
platformdirs==3.8.0
pluggy==1.2.0
pre-commit==2.21.0
prometheus-client==0.17.0
prompt-toolkit==3.0.38
protobuf==3.19.6
psutil==5.9.5
pyasn1==0.5.0
pyasn1-modules==0.3.0
pycparser==2.21
pydata-sphinx-theme==0.13.3
PyGithub==1.59.0
Pygments==2.15.1
PyJWT==2.7.0
pykalman==0.9.5
PyNaCl==1.5.0
pynwb==2.3.3
pyparsing==3.1.0
pyreadline==2.1
pyrsistent==0.19.3
PySide2==5.14.1
pytest==7.4.0
pytest-cov==3.0.0
pytest-qt==4.2.0
pytest-xvfb==3.0.0
python-dateutil @ file:///tmp/build/80754af9/python-dateutil_1626374649649/work
python-rapidjson==1.10
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1680088766131/work
PyVirtualDisplay==3.0
PyWavelets==1.3.0
pywin32==306
pywin32-ctypes==0.2.1
pywinpty==1.1.6
PyYAML==6.0
pyzmq==25.1.0
qimage2ndarray==1.10.0
QtPy @ file:///home/conda/feedstock_root/build_artifacts/qtpy_1680148448366/work
qudida==0.0.4
readme-renderer==37.3
regex==2023.6.3
requests==2.31.0
requests-oauthlib==1.3.1
requests-toolbelt==1.0.0
rfc3986==2.0.0
rich==10.16.1
rsa==4.9
ruamel.yaml==0.17.32
ruamel.yaml.clib==0.2.7
scikit-image==0.19.3
scikit-learn==1.0.2
scikit-video==1.1.11
scipy @ file:///C:/bld/scipy_1637806996411/work
seaborn==0.12.2
segmentation-models==1.0.1
Send2Trash==1.8.2
Shapely==1.7.1
shiboken2==5.14.1
six @ file:///tmp/build/80754af9/six_1644875935023/work
sleap==1.4.1
smmap==5.0.0
sniffio==1.3.0
snowballstemmer==2.2.0
soupsieve==2.4.1
Sphinx==4.5.0
sphinx-autobuild==2021.3.14
sphinx-book-theme==1.0.1
sphinx-copybutton==0.5.2
sphinx-togglebutton==0.3.2
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
SQLAlchemy==1.4.48
tensorboard==2.8.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow==2.8.4
tensorflow-estimator==2.8.0
tensorflow-hub==0.13.0
tensorflow-io-gcs-filesystem==0.31.0
termcolor==2.3.0
terminado==0.17.1
threadpoolctl==3.1.0
tifffile==2021.11.2
tinycss2==1.2.1
toml==0.10.2
tomli==2.0.1
tornado==6.2
tqdm==4.65.0
traitlets==5.9.0
twine==3.3.0
typed-ast==1.5.4
typing_extensions==4.6.3
tzdata==2023.3
tzlocal==5.0.1
uc-micro-py==1.0.2
urllib3==1.26.16
virtualenv==20.23.1
wcwidth==0.2.6
webencodings==0.5.1
websocket-client==1.6.1
Werkzeug==2.2.3
widgetsnbextension==3.6.4
wrapt==1.15.0
zipp==3.15.0
</code></pre>
</details>

<details><summary>Logs</summary>
<pre><code># Conda/mamba install attempt
mamba create -n sleap -c conda-forge -c nvidia -c sleap/label/dev -c sleap -c anaconda sleap=1.4.1
Error: Could not solve for environment specs — no viable option for tensorflow==2.7.0

# Pip install attempt with Python >=3.9
mamba create -n sleap python=3.9
mamba activate sleap
pip install sleap[pypi]==1.4.1
ERROR: No matching distribution found for PySide2<=5.14.1,>=5.13.2
</code></pre>
</details>

<h2>Reproduction Steps</h2>
<ol>
<li>Attempt conda/mamba installation (fails):
<pre><code>mamba create -n sleap -c conda-forge -c nvidia -c sleap/label/dev -c sleap -c anaconda sleap=1.4.1</code></pre>
</li>
<li>Attempt pip installation on Python ≥3.9 (fails):
<pre><code>mamba create -n sleap python=3.9
mamba activate sleap
pip install sleap[pypi]==1.4.1</code></pre>
</li>
<li>Successful workaround using Python 3.8:
<pre><code>mamba create -n sleap python=3.8
mamba activate sleap
pip install sleap[pypi]==1.4.1</code></pre>
</li>
</ol>

<p><strong>Links for reference:</strong></p>
<ul>
<li>TensorFlow 2.7.0 PyPI wheels: <a href="https://pypi.org/project/tensorflow/2.7.0/#files">https://pypi.org/project/tensorflow/2.7.0/#files</a></li>
<li>PySide2 PyPI wheels: <a href="https://pypi.org/project/PySide2/#files">https://pypi.org/project/PySide2/#files</a></li>
<li>Conda-forge TF 2.7.0 (no Windows builds): <a href="https://anaconda.org/conda-forge/tensorflow/files?version=2.7.0">https://anaconda.org/conda-forge/tensorflow/files?version=2.7.0</a></li>
<li>Related GitHub issue: <a href="https://github.com/talmolab/sleap/issues/2126">https://github.com/talmolab/sleap/issues/2126</a></li>
</ul>

<h2>Suggested Actions</h2>
<ol>
<li>Provide Windows-compatible TensorFlow 2.7.0 builds in the SLEAP conda channel or update dependency to a TF version available on conda-forge/anaconda for Windows.</li>
<li>Update PySide2 dependency pin to a version with wheels for Python ≥3.9, or clearly document Python 3.8 as a requirement for GUI installation.</li>
<li>Include a tested pip install workflow for Windows in the official installation instructions.</li>
</ol>

**Comment by gitttt-1234** [comment]
Hi @GarrettBlair !

Thank you for reporting this!

(We are fully switching over to a PyTorch backend [SLEAP-NN](https://github.com/talmolab/sleap-nn) very soon, which should simplify and resolve the dependency issues.)

Thanks,
Divya

**Comment by alicup29** [comment]
Hi @GarrettBlair !

The TensorFlow and PySide2 dependency issues you ran into have been resolved w/ SLEAP v1.5+ migrating to a PyTorch backend ([via `sleap-nn`](https://docs.sleap.ai/latest/guides/migrating-to-sleap-1-5/#uv-based-installation)) and switching from PySide2 to PySide6 for the GUI.

To install the latest version on Windows:
```
uv tool install --python 3.13 "sleap[nn]==1.6.3" --with "sleap-io==0.7.0" --with "sleap-nn==0.2.0" --torch-backend auto
```

and the full installation docs: https://docs.sleap.ai/latest/installation/

Closing this as resolved for now, but if you run into any issues, please open a new issue and we'll help out!

Thanks!
Amick
